topdown/cache/cache.go

```diff
@@ -67,24 +98,40 @@ type InterQueryCacheValue interface {
 // InterQueryCache defines the interface for the inter-query cache.
 type InterQueryCache interface {
 	Get(key ast.Value) (value InterQueryCacheValue, found bool)
-	Insert(key ast.Value, value InterQueryCacheValue) int
+	Insert(key ast.Value, value InterQueryCacheValue, expiresAt time.Time) int
 	Delete(key ast.Value)
 	UpdateConfig(config *Config)
 	Clone(value InterQueryCacheValue) (InterQueryCacheValue, error)
 }
 
 // NewInterQueryCache returns a new inter-query cache.
 func NewInterQueryCache(config *Config) InterQueryCache {
-	return &cache{
+	iqCache := &cache{
 		items:  map[string]cacheItem{},
 		usage:  0,
 		config: config,
 		l:      list.New(),
 	}
+
+	// Start routine to clean up stale values once every StaleEntryEvictionPeriodSeconds
+	cleanupPeriod := iqCache.staleEntryEvictionTimePeriodSeconds()
+	if cleanupPeriod > 0 {
+		ticker := time.NewTicker(time.Duration(cleanupPeriod) * time.Second)
```

Comment:
is it okay to skip graceful shutdown here and keep the timer ticking? The `for {}` loop below will run (block shutdown) indefinitely, won't it?

Reply:
Good point. Let me see if I can get context from callers and handle `<-ctx.Done` here. I think that should handle graceful shutdown.

Reply:
@lukyer Good point. Let me see if I can get context from callers and handle `<-ctx.Done` here. I think that should handle graceful shutdown.

Reply:
Graceful shutdown with log to indicate ticker has stopped:

![image](https://github.com/open-policy-agent/opa/assets/20646660/88a7d922-5472-477c-b55a-344085fced00)